### DISCUSSION TITLE
Databricks Error SparkLinker

### DISCUSSION THREAD

**Discussion by Mac-delValle** [moj-analytical-services/splink#1461] [Q&A]
Hi all,

I am new to databricks and spark/python and very excited to use Splink. I am trying to run Splink in databricks. I am getting an error 'NoneType' object has no attribute 'sparkContext' when trying to execute the command linker = SparkLinker(df, settings). I am using the default settings provided in the example repo. Any help is much appreciated. Thank you in advance!

Here are my cluster configurations:

- 10.4 LTS (includes Apache Spark 3.2.1, Scala 2.12)

When I did the pip install, I realized the version has been updated since the demo repo was last published so I specified version 3.9.3.

Here is the full error message provided by databricks:

AttributeError                            Traceback (most recent call last)
<command-689305410732898> in <module>
      1 # from splink.spark.linker import SparkLinker
----> 2 linker = SparkLinker(df, settings)
      3 deterministic_rules = [
      4     "l.first_name = r.first_name and levenshtein(r.dob, l.dob) <= 1",
      5     "l.surname = r.surname and levenshtein(r.dob, l.dob) <= 1",

/local_disk0/.ephemeral_nfs/envs/pythonEnv-xxx-xx-xxxxxx-xxxx/lib/python3.8/site-packages/splink/spark/linker.py in __init__(self, input_table_or_tables, settings_dict, break_lineage_method, set_up_basic_logging, input_table_aliases, spark, validate_settings, catalog, database, repartition_after_blocking, num_partitions_on_repartition)
    182         self.in_databricks = "DATABRICKS_RUNTIME_VERSION" in os.environ
    183         if self.in_databricks:
--> 184             enable_splink(spark)
    185 
    186         self._set_default_break_lineage_method()

/local_disk0/.ephemeral_nfs/envs/pythonEnv-xxx-xx-xxxxxx-xxxx/lib/python3.8/site-packages/splink/databricks/enable_splink.py in enable_splink(spark)
     13         None
     14     """
---> 15     sc = spark.sparkContext

**Comment by RobinL** [comment]
Hiya,

The error is being created from:
https://github.com/moj-analytical-services/splink/blob/31d4d0ec9e5c7f27abc25e57968997841a8d004c/splink/databricks/enable_splink.py#L15

which is run automatically for you since you're in a databricks environment, see [here](https://github.com/moj-analytical-services/splink/blob/31d4d0ec9e5c7f27abc25e57968997841a8d004c/splink/spark/linker.py#L184)

The spark context is obtained from your input Spark dataframes using [this function](https://github.com/moj-analytical-services/splink/blob/31d4d0ec9e5c7f27abc25e57968997841a8d004c/splink/spark/linker.py#L147)

Specifically:
`self.spark = t.sql_ctx.sparkSession`

So, things for you to check:

What happens if you run `df.sql_ctx.sparkSession`.  Do you get a SparkSession back. If not , what do you get

Another option is to pass Spark explicitly into the SparkLinker see [here](https://github.com/moj-analytical-services/splink/blob/31d4d0ec9e5c7f27abc25e57968997841a8d004c/splink/spark/linker.py#L117), which I think should avoid the problem.

I'm afraid I don't have access to databricks so I'm unable to check this easily.

**Reply by sazarela** [reply]
I just ran into this issue as well. I'm extremely new to this. I don't completely understand what the solution is based on this. 

EDIT: For those looking at this.. I fixed the error by changing the DBR from 13.3 LTS to 12.2 LTS ML

**Reply by beedrumms** [reply]
Hi! I have looking into this problem (as i experienced it myself) and essentially have determined that SparkLinker fails when it calls enable_splink() in a databricks environment, specifically where it is trying to confirm registration of Splink on your Databricks cluster. Unfortunately, passing Spark into the SparkLinker (or enable_spark) does not solve the issue because i think the root of the problem is cluster security features. Clusters w/ a runtime over 13.2 or clusters in shared access mode automatically have an isolation feature that limits users' writing permissions. 

If you run the following code in a cluster rt over 13.2 or a shared access cluster, you would probably get the same error: 
`from splink.spark.jar_location import similarity_jar_location`
`sc = spark.sparkContext`
`_jar_path = similarity_jar_location()` # returns path for scala-udf.jar
`JavaURI = sc._jvm.java.net.URI` # pyfj.java_gateway JavaClass  
`JavaURI.create("file:" + _jar_path)` ## the jarpath will be the path you shared above ### Code should fail here in non-compatible cluster rts or access modes. 

However, the above code works perfect in a cluster rt of 13.2 (now depreciated) and below, and it should return a hash num. 

All this is leading up to that I dont think that there is anything to do about this rn except for whitelisting java classes (or changing enable_splink?), however that comes w/ risk! Your solution of downgrading rt appears to be the best solution atm! Hope this was sort of helpful as i am working through this as well :)

**Reply by RobinL** [reply]
Hi @beedrumms  - that's super useful.  I'm not a databricks user so I have no feel for the best solution to this.  I don't suppose you know/could recommend what a better approach could be?

A couple of thoughts/questions:
- The jars are [here](https://github.com/moj-analytical-services/splink/tree/master/splink/files/spark_jars).  Are you able to manually register them somehow, or does that also fail the whitelist?
- If we were to (somehow) make the jars available Maven, would that help?

Or is there a better way you can think of?

p.s. depending on the size of your data, you may have noticed we recently [changed the recommended](https://moj-analytical-services.github.io/splink/topic_guides/splink_fundamentals/backends/backends.html?h=backends#duckdb) backend to duckdb for most data, assuming you have access to a reasonably high-spec system.  This was off the back of these benchmarking results of the recent 3.9.11 release which dramatically improved the speed of duckdb workflows https://www.robinlinacre.com/fast_deduplication/.  

Also to note: there are no plans to drop support for Spark, so no need to worry if that's your preference

**Reply by sthamodh** [reply]
Hi @RobinL and @beedrumms,

I can confirm a couple of things from above:

1. I have used 12.2 LTS when testing the splink solution on a compute cluster with a service principal and it worked without any errors
2. When I tried to use 13.3 LTS, it yielded me errors
3. Specifically, blacklist errors were raised when the splink libary tried to create a spark context with the scalaudfs from splink

```
import splink
from splink.spark.jar_location import similarity_jar_location
conf = SparkConf()
conf.set("spark.driver.memory", "32g")
conf.set("spark.default.parallelism", "32")
path = similarity_jar_location()
conf.set("spark.jars", path)

sc = SparkContext.getOrCreate(conf=conf)
```

When our organization spoke to databricks about this, we were able to confirm that the issue here is that Scala (and other lower level libraries) run directly on the JVM and isolation between a given user's data and another user's data segment in memory cannot be guaranteed. This is because the notebook that houses the splink code is running on a compute cluster that is shared between multiple users or has passthrough enabled.

Two of the following options worked for me:

1. Use a no-isolation shared cluster with a service principal when such a cluster is meant to serve multiple users <OR>
2. Use a single user cluster i.e. each user gets their own personal cluster with 12.2 LTS

As for your suggestion @RobinL, I believe one can install jars from Maven according to this: https://learn.microsoft.com/en-us/azure/databricks/libraries/package-repositories

Hope this helps!

**Reply by jericksonclinicaloptions** [reply]
My understanding is you must use runtime 12.2 LTS (I use the ML variant). 
To install the jar file in Databricks, I downloaded it locally and then installed to DBFS on the cluster:
![image](https://github.com/moj-analytical-services/splink/assets/104126080/c72716f8-4169-4553-8c76-916e0ae769e8)
Also, you can set the spark configuration under **Advance options** on the cluster since Databricks automatically sets up the spark context when the cluster starts:
![image](https://github.com/moj-analytical-services/splink/assets/104126080/5601d4b5-3d66-41e8-a80d-b0c94dba5a6b)

**Reply by beedrumms** [reply]
@RobinL  apologies for the delay and @sthamodh thanks for the response! Unfortunately, manual registration of the jar did not work, but there is small progress: 

the 'enable_splink' module used to fail when it uses java method to write to the following location:  
       file:/local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/splink/files/spark_jars/scala-udf-similarity-0.1.1_spark3.x.jar

**Without** the jar manually registered, the module would break at `JarURI = JavaURI.create("file:" + _jar_path)` 

 Now **with the scala-udf jar manually registered on cluster** via dbfs, new error point is just below, at: 

`JavaJarId(JarURI,
                ManagedLibraryId.defaultOrganization(),
                NoVersionModule.simpleString()
                )`

The new error is below:
![image](https://github.com/moj-analytical-services/splink/assets/85265112/d5773a72-26f4-419c-b18e-a74a1dab1d56)

This was all **tested between a 12.2 RT and a 14.0 RT (as 13.3 is depreciated)**. Parallelism and partition configurations have no baring on the outcome. I have been discussing options w/ our data eng team w/ respect to writing the jar to the unity catalog or azure blob storage.

**Reply by RobinL** [reply]
@beedrumms. thanks very much for the additional info. Apologies if you'd already seen it, but we recently had an external contribution (not yet released, but on `master`) which fixes this for some users.  I wondered if that helps at all for you?

See here: https://github.com/moj-analytical-services/splink/pull/1973

**Reply by beedrumms** [reply]
@RobinL thanks so much for this -- this code looks like it would have worked perfectly to address the RT issue -- However, we did solve the RT issue by:

1) uploading splink as a .whl file to a shared databricks workspace 
2) then installing the library on our cluster by referencing the workspace location, 

This was tested on a runtime of 14.2 and worked perfectly. I discovered this while coming into compliance w/ databricks depreciation of init scripts in dbfs so the [azure docs](https://learn.microsoft.com/en-us/azure/databricks/_extras/documents/azure-init-workspace-files.pdf) to use a workspace for library was helpful in implementing this solution. I will note that this does not solve the shared access cluster compatibility issue -- this seems to be a whole other can of worms, but for now, its working great for us! thanks so much
